llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,323 @@
+  /// If the new size is smaller than the current size, the elements of the
+  /// pages that are not needed anymore will be destroyed, however, elements of
+  /// the last page will not be destroyed.
+  ///
+  /// For these reason the usage of this vector is discouraged if you rely
+  /// on the construction / destructor of the elements to be invoked.
+  void resize(size_t NewSize) {
+    if (NewSize == 0) {
+      clear();
+      return;
+    }
+    // Handle shrink case: destroy the elements in the pages that are not
+    // needed anymore and deallocate the pages.
+    //
+    // On the other hand, we do not destroy the extra elements in the last page,
+    // because we might need them later and the logic is simpler if we do not
+    // destroy them. This means that elements are only destroyed only when the
+    // page they belong to is destroyed. This is similar to what happens on
+    // access of the elements of a page, where all the elements of the page are
+    // constructed not only the one effectively neeeded.
+    if (NewSize < Size) {
+      size_t NewLastPage = (NewSize - 1) / PageSize;
+      for (size_t I = NewLastPage + 1, E = PageToDataPtrs.size(); I < E; ++I) {
+        T *PagePtr = PageToDataPtrs[I];
+        if (PagePtr == InvalidPage)
+          continue;
+        T *Page = PagePtr;
+        // We need to invoke the destructor on all the elements of the page.
+        for (size_t J = 0; J < PageSize; ++J)
+          Page[J].~T();
+        Allocator.getPointer()->Deallocate(Page);
+        // We mark the page invalid, to avoid double deletion.
+        PageToDataPtrs[I] = InvalidPage;
+      }
+      PageToDataPtrs.resize(NewLastPage + 1);
+    }
+    Size = NewSize;
+    // If the capacity is enough, just update the size and continue
+    // with the currently allocated pages. Notice that we do not
+    // need to default construct any new element, because that was already done
+    // when the page was allocated.
+    if (Size <= capacity())
+      return;
+    // The number of pages to allocate. The Remainder is calculated
+    // for the case in which the NewSize is not a multiple of PageSize.
+    // In that case we need one more page.
+    size_t Pages = Size / PageSize;
+    size_t Remainder = Size % PageSize;
+    if (Remainder != 0)
+      Pages += 1;
+    assert(Pages > PageToDataPtrs.size());
+    // We use InvalidPage to indicate that a page has not been allocated yet.
+    // This cannot be 0, because 0 is a valid page id.
+    // We use InvalidPage instead of a separate bool to avoid wasting space.
+    PageToDataPtrs.resize(Pages, InvalidPage);
+  }
+
+  [[nodiscard]] bool empty() const { return Size == 0; }
+
+  /// Clear the vector, i.e. clear the allocated pages, the whole page
+  /// lookup index and reset the size.
+  void clear() {
+    Size = 0;
+    for (T *Page : PageToDataPtrs) {
+      if (Page == InvalidPage)
+        continue;
+      for (size_t I = 0; I < PageSize; ++I)
+        Page[I].~T();
```

Comment:
```suggestion
      std::destroy_n(Page, PAGE_SIZE);
```
per https://github.com/llvm/llvm-project/pull/66430#discussion_r1330548934